tests/rbd_mirror/test_delayed_replication_handler.py

```diff
@@ -0,0 +1,190 @@
+        )
+        with parallel() as p:
+            p.spawn(mirror2.wait_for_replay_complete, imagespecs[0])
+            p.spawn(mirror2.wait_for_replay_complete, imagespecs[1])
+
+        return export_and_get_md5sum()
+
+    def write_data():
+        with parallel() as p:
+            for imagespec in imagespecs:
+                p.spawn(mirror1.benchwrite, imagespec=imagespec, io="100M")
+
+    def write_data_and_verify_no_mirror_till_delay():
+        starttime = datetime.datetime.now()
+        wait_till = datetime.timedelta(seconds=(delay - 5))
+        log.info(f"{starttime}")
+        write_data()
+        while datetime.datetime.now() - starttime <= wait_till:
+            if initial_md5sums != export_and_get_md5sum():
+                log.error("Mirorring activity seen on secondary without desired delay")
+                return 1
+            log.debug(
+                "Verified that no data has bee added to secondary image in last 50 seconds"
+            )
+            time.sleep(50)
+
+        log.info(
+            "No mirorring ios seen in secondary untill the delay, Test case passed"
+        )
+        return 0
+
+    def config_delay_at_secondary():
+        mirror2.exec_cmd(
+            cmd=f"ceph config set client rbd_mirroring_replay_delay {delay}"
+        )
+
+    try:
+        if not config.get("skip_initial_config"):
+            rbd_mirror_config(**kw)
+
+        initial_md5sums = get_initial_md5sums()
+
+        if config.get("delay_at_secondary"):
+            log.debug(
+                "Executing delayed replication scenario when delay configured at secondary"
+            )
+            config_delay_at_secondary()
+            return write_data_and_verify_no_mirror_till_delay()
+
+        if config.get("direct_failover_with_delay"):
+            config_delay_at_secondary()
+            time.sleep(20)
+
+            write_data()
+            time.sleep(20)
+            with parallel() as p:
+                for imagespec in imagespecs:
+                    p.spawn(mirror2.promote, force=True, imagespec=imagespec)
+            time.sleep(20)
+            if initial_md5sums != export_and_get_md5sum():
+                log.error("Mirroring has happened during failover unexpectedly")
+                return 1
+            else:
+                log.debug(
+                    "Delay has been respected during Image force promote, TC passed"
+                )
+                return 0
+
+    finally:
+        if config.get("delay_at_primary"):
+            for imagespec in imagespecs:
+                rbd1.image_meta(
+                    action="set",
+                    image_spec=imagespec,
+                    key="conf_rbd_mirroring_replay_delay",
+                    value=0,
+                )
+        else:
+            mirror2.exec_cmd(cmd="ceph config set client rbd_mirroring_replay_delay 0")
+        time.sleep(10)
```

Comment:
why we need to sleep at end?

Reply:
Wanted to give space before execution of upcoming TC so that the execution of upcoming TC doesn't get affected by delay unnecessarily.
(new TCs that might get added after these TCs)